sdk/program/src/vote/state/vote_state_deserialize.rs

```diff
@@ -1,36 +1,68 @@
 use {
+    super::{MAX_EPOCH_CREDITS_HISTORY, MAX_LOCKOUT_HISTORY},
     crate::{
+        clock::Epoch,
         instruction::InstructionError,
+        pubkey::Pubkey,
         serialize_utils::cursor::*,
-        vote::state::{BlockTimestamp, LandedVote, Lockout, VoteState, MAX_ITEMS},
+        vote::{
+            authorized_voters::AuthorizedVoters,
+            state::{BlockTimestamp, LandedVote, Lockout, VoteState, MAX_ITEMS},
+        },
     },
-    std::io::Cursor,
+    std::{collections::VecDeque, io::Cursor, ptr::addr_of_mut},
 };
 
 pub(super) fn deserialize_vote_state_into(
     cursor: &mut Cursor<&[u8]>,
-    vote_state: &mut VoteState,
+    vote_state: *mut VoteState,
     has_latency: bool,
 ) -> Result<(), InstructionError> {
-    vote_state.node_pubkey = read_pubkey(cursor)?;
-    vote_state.authorized_withdrawer = read_pubkey(cursor)?;
-    vote_state.commission = read_u8(cursor)?;
-    read_votes_into(cursor, vote_state, has_latency)?;
-    vote_state.root_slot = read_option_u64(cursor)?;
-    read_authorized_voters_into(cursor, vote_state)?;
+    // General safety note: we must use add_or_mut! to access the `vote_state` fields as the value
+    // is assumed to be _uninitialized_, so creating references to the state or any of its inner
+    // fields is UB.
+
+    read_pubkey_into(
+        cursor,
+        // Safety: if vote_state is non-null, node_pubkey is guaranteed to be valid too
+        unsafe { addr_of_mut!((*vote_state).node_pubkey) },
+    )?;
+    read_pubkey_into(
+        cursor,
+        // Safety: if vote_state is non-null, authorized_withdrawer is guaranteed to be valid too
+        unsafe { addr_of_mut!((*vote_state).authorized_withdrawer) },
+    )?;
+    let commission = read_u8(cursor)?;
+    let votes = read_votes(cursor, has_latency)?;
+    let root_slot = read_option_u64(cursor)?;
+    let authorized_voters = read_authorized_voters(cursor)?;
     read_prior_voters_into(cursor, vote_state)?;
-    read_epoch_credits_into(cursor, vote_state)?;
+    let epoch_credits = read_epoch_credits(cursor)?;
     read_last_timestamp_into(cursor, vote_state)?;
 
+    // Safety: if vote_state is non-null, all the fields are guaranteed to be
+    // valid pointers.
+    //
+    // Heap allocated collections - votes, authorized_voters and epoch_credits -
+    // are guaranteed not to leak after this point as the VoteState is fully
+    // initialized and will be regularly dropped.
+    unsafe {
+        addr_of_mut!((*vote_state).commission).write(commission);
+        addr_of_mut!((*vote_state).votes).write(votes);
+        addr_of_mut!((*vote_state).root_slot).write(root_slot);
+        addr_of_mut!((*vote_state).authorized_voters).write(authorized_voters);
+        addr_of_mut!((*vote_state).epoch_credits).write(epoch_credits);
+    }
```

Comment:
hmm, i can assume this deserialization can be run on untrusted data, right? (i.e. it's fallible as coded with `?`s).

in that case, i think these fields need to be initialized in the case of errors (i.e. when bailing out with one of those above `?`s hit by `InvalidAccountData`).

Otherwise, `std::ptr::drop_in_place()`-ed `vote_state: &mut VoteState` will leave some fields uninitialzied when returning `Err(_)` from `deserialize_into()`. And it will cause uaf eventually when the referenced `VoteState` is dropped sometime later, right?

On the other hand, `deserialize_into_uninit(vote_state: &mut MaybeUninit<VoteState>)` should be safe, assuming `assume_init()` won't called on `vote_state` at the caller when `Err(_)` is returned from this `fn`.

Reply:
considering `deserialize_into` impl will be deprecated and short-lived, how about just restoring old impl instead of flowing into the shiny `deserialize_into_uninit()`?

Reply:
we'd have to duplicate all the code, since I moved the old code to working with `*mut VoteState`. I think there's a better fix, let me do that and add a test.

Reply:
I've pushed a fix for this. When deserialize_into_ptr fails, we reset to `VoteState::default()`